proposals/stdlib/random.md

```diff
@@ -0,0 +1,253 @@
+* [Apache RNG](https://commons.apache.org/proper/commons-rng/userguide/rng.html): 
+[UniformRandomProvider](https://commons.apache.org/proper/commons-rng/commons-rng-client-api/apidocs/org/apache/commons/rng/UniformRandomProvider.html)
+
+## Motivation and use cases
+
+While there's no problem with generating random numbers in Kotlin/JVM (aside that `ThreadLocalRandom`
+isn't available prior to JDK 7), there's no API yet to do that in Kotlin for other targets, 
+and more importantly there's no way to do that consistently in multiplatform code.
+
+The use cases of random API are established, but here we name a few:
+
+* Sampling a random element from a collection
+* Collection shuffling
+* Random game content and level generation
+* Simulations requiring a source of random numbers
+* Using randomness in tests to cover more code execution paths
+
+## Alternatives
+
+The current alternatives require finding some platform libraries and providing common multiplatform API
+for them or implementing random generators by hand. 
+
+## Placement
+
+This API shall be placed into the Kotlin Standard Library. Regarding the package the following options were considered:
+
+- Use the existing `kotlin` package which is imported by default.
+- Use the existing `kotlin.math` package
+- Create new `kotlin.random` package and place API there.
+
+We decided to stick with the latter as it would be more explorable in docs: all the related API will be shown close to each other.
+Also it's more future proof in regard to possible API additions.
+
+A downside of a new package is that it will require new imports, especially if there will be extensions.
+A class named `Random` from that package can be confused with `java.util.Random` in the completion list.
+It should be investigated whether we could make that package imported by default in Kotlin 1.3.
+
+## Reference implementation
+
+The initial implementation is [available](https://github.com/JetBrains/kotlin/compare/4d51d13~1...f7337cc) in 1.3-M1 branch.
+
+## Dependencies
+
+What are the dependencies of the proposed API:
+
+* JDK: `java.util.Random` in JDK 6,7, `ThreadLocalRandom` in JDK 8+
+* JS: `Math.random` function as the source of randomness seed
+
+## API details
+
+The following code snippet summarizes the public API introduced by this proposal.
+
+```kotlin
+abstract class Random {
+
+    abstract fun nextBits(bitCount: Int): Int
+
+    open fun nextInt(): Int 
+    open fun nextInt(bound: Int): Int 
+    open fun nextInt(origin: Int, bound: Int): Int
+    open fun nextInt(range: IntRange): Int
+
+    open fun nextLong(): Long
+    open fun nextLong(bound: Long): Long
+    open fun nextLong(origin: Long, bound: Long): Long
+    open fun nextLong(range: LongRange): Long
+    
+    open fun nextBoolean(): Boolean
+   
+    open fun nextDouble(): Double
+    open fun nextDouble(bound: Double): Double
+    open fun nextDouble(origin: Double, bound: Double): Double
```

Comment:
What was the reason to have 3 overloaded methods instead of one with default parameters like `fun nextDouble(origin: Double = 0.0, bound: Double = Double.MAX_VALUE)`?

Reply:
- `nextDouble()` returns value from 0.0 and up to 1.0
- overloads with less parameters allow more efficient implementations (less parameter checking, less edge cases)

Reply:
I see. Then 2 questions:
 1) How much more efficient is `nextDouble()` than a method with default parameters?
 2) Is that efficiency worth having a separate `nextDouble()`? 
 3) Is that efficiency still worth having a relatively rare `nextDouble(bound: Double)`? 

Note that 2) and 3) combined effectively bloat the API two fold.

Reply:
I don't have the precise measurements, but all of the following operations definitely add up. `nextDouble` without parameters doesn't have to:
- check if the optional parameters are specified and substitute them with their default values
- check if those parameters are valid
- calculate the range `bound - origin`
- check if it is finite
- multiply the random [0, 1) by the range and add the origin
- check if the resulting value is equal to bound.